data_science/palette/notebooks/00 - RGB, LAB, and human colour perception.ipynb

```diff
@@ -0,0 +1,133 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def colour_distance(colour_1, colour_2):\n",
+    "    r_1, g_1, b_1 = colour_1\n",
+    "    r_2, g_2, b_2 = colour_2\n",
+    "\n",
+    "    return (2 * (r_1 - r_2) ** 2 +\n",
+    "            4 * (g_1 - g_2) ** 2 +\n",
+    "            3 * (b_1 - b_2) ** 2) ** 0.5"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can improve further by adding some extra weirdness to the red and blue channels at the end\n",
+    "\n",
+    "${\\displaystyle \\Delta C={\\sqrt {2\\times \\Delta R^{2}+4\\times \\Delta G^{2}+3\\times \\Delta B^{2}+{{{\\bar {r}}\\times (\\Delta R^{2}-\\Delta B^{2})} \\over {256}}}}}$\n",
+    "\n",
+    "Here it is in python"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def colour_distance(colour_1, colour_2):\n",
+    "    r_1, g_1, b_1 = colour_1\n",
+    "    r_2, g_2, b_2 = colour_2\n",
+    "    \n",
+    "    d_r_sq = (r_1 - r_2) ** 2\n",
+    "    d_g_sq = (g_1 - g_2) ** 2\n",
+    "    d_b_sq = (b_1 - b_2) ** 2\n",
+    "    mean_r = (r_1 + r_2) / 2\n",
+    "    \n",
+    "    d_c_sq = (2 * d_r_sq + 4 * d_g_sq + 3 * d_b_sq +\n",
+    "              (mean_r * (d_r_sq - d_b_sq) / 256))\n",
+    "\n",
+    "    return d_c_sq ** 0.5"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The most general and efficient approach (as far as I know) is to transform the entire image's RGB coordinates into a new space. The CIE produced [CIELAB](https://en.wikipedia.org/wiki/CIELAB_color_space#CIELAB) to better approximate human perception. The axes of the space are as follows\n",
+    "\n",
+    "> The three coordinates of CIELAB represent the lightness of the color (`L` = 0 yields black and `L` = 100 indicates diffuse white; specular white may be higher), its position between red/magenta and green (`a`, negative values indicate green while positive values indicate magenta) and its position between yellow and blue (`b`, negative values indicate blue and positive values indicate yellow).\n",
+    "\n",
+    "[CIE76](https://en.wikipedia.org/wiki/Color_difference#CIE76) (ie euclidian distance in LAB space) was the original distance proposed with the space. It's been improved upon since, but the differences are minor and as far as I've seen, are unnecessary complications for such minimal gain.\n",
+    "\n",
+    "Implementing the map from RGB to LAB and back is incredibly simple - all we need to do is import the relevant function from `skimage`"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from skimage.color import rgb2lab, lab2rgb"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can now use our original, super-simple colour distance!"
```

Comment:
How? An example would be helpful. I think I know what I’d do – convert RGB to LAB, then back. Helper methods would be useful:

```python
def rgb_colour_distance(colour_1, colour_2):
    lab_colour_1 = rgb2lab(colour_1)
    lab_colour_2 = rgb2lab(colour_2)
    return lab_colour_distance(lab_colour_1, lab_colour_2)

def lab_colour_distance(colour_1, colour_2):
    return euclidean_distance(colour_1, colour_2)
```